src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1111,11 +1111,7 @@ HeapWord* ShenandoahHeap::allocate_from_plab_slow(Thread* thread, size_t size, b
   }
   future_size = MIN2(future_size, PLAB::max_size());
   future_size = MAX2(future_size, PLAB::min_size());
-
-  size_t unalignment = future_size % CardTable::card_size_in_words();
-  if (unalignment != 0) {
-    future_size = future_size - unalignment + CardTable::card_size_in_words();
-  }
+  future_size = align_down(future_size, CardTable::card_size_in_words());
```

Comment:
Can one use cardsize-aligned min and max sizes (taking whetever PLAB gives us, may be by using a ShenandoahPLAB subclass. We then first align future_size up wrt card size, then floor/clamp it with the already cardsize-aligned min and max?

Reply:
We also need a comment somewhere stating why we need these to be card-aligned.

I wonder also if this is entirely correct in itself, unless the shared allocation itself also always leaves us with a card-aligned top from which these PLABs are allocated.

In particular, it would be good to assert that `allocate_new_plab()` always gives us back a card-aligned start address.

Would be good to also use `align_up` or `align_down` in `allocate_new_plab()` as well. Indeed, it might be a good idea to do so in any code in Shenandoah where we are unnecessarily rolling our own custom alignment code.

Reply:
There's also a couple of assertion checks in `ShenandoahFreeSet::allocate_aligned_plab()` that use the modulus operation, which should probably use `is_aligned()` instead.

Would be good to clean those up as well while we are taking care of this.

Reply:
Thanks for these suggestions.  I believe I have addressed each in the newest rev of the code.